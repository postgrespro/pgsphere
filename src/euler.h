#ifndef __PGS_EULER_H__
#define __PGS_EULER_H__

#include "point.h"

/*
 * Euler transformation declarations
 */

/*
 * Data structure of spherical (Euler) transformation.
 */
typedef struct
{
	unsigned char	phi_a:2,	/* first axis */
					theta_a:2,	/* second axis */
					psi_a:2;	/* third axis */
	float8			phi,		/* first rotation angle */
					theta,		/* second rotation angle */
					psi;		/* third rotation angle */
} SEuler;


/*
 * Transforms a spherical point and returns the pointer to a transformed spherical
 * point.
 */
void	euler_spoint_trans(SPoint *out, const SPoint *in, const SEuler *se);

/*
 * Transforms a spherical vector from 'spb' to 'spe' into an Euler transformation.
 * Returns true if the transformation was successful.
 */
bool	spherevector_to_euler(SEuler *se, const SPoint *spb, const SPoint *spe);

/*
 * Sets the axes of transformation to ZXZ.
 */
void	seuler_set_zxz(SEuler *se);

/*
 * Checks equality of two transformations.
 */
bool	strans_eq(const SEuler *e1, const SEuler *e2);

/*
 * Transforms a vector using an Euler transformation. Returns the pointer to
 * the result vector.
 */
void	euler_vector_trans(Vector3D *out, const Vector3D *in, const SEuler *se);

/*
 * Inverts an Euler transformation. Returns the pointer to the
 * inverted transformation.
 */
void	spheretrans_inverse(SEuler *se_out, const SEuler *se_in);

/*
 * Inverts an Euler transformation replacing the original Euler transformation.
 * Returns the pointer to the inverted transformation.
 */
void	spheretrans_inv(SEuler *se);

/*
 * Converts an Euler transformation to a ZXZ-axis transformation. Returns
 * the pointer to the converted transformation.
 */
void	strans_zxz(SEuler *ret, const SEuler *se);

/*
 * Transforms an Euler transformation 'in' into 'out' using 'se'. The result
 * is always a ZXZ-axis transformation. Returns the pointer to the transformed
 * transformation.
 */
void	seuler_trans_zxz(SEuler *out, const SEuler *in, const SEuler *se);

/*
 * Input of an Euler transformation.
 */
Datum	spheretrans_in(PG_FUNCTION_ARGS);

/*
 * Input of an Euler transformation with axis Z,X,Z from three angles
 * (phi, theta, psi) in radians.
 */
Datum	spheretrans_from_float8(PG_FUNCTION_ARGS);

/*
 * Returns the first angle of an Euler transformation in radians.
 */
Datum	spheretrans_phi(PG_FUNCTION_ARGS);

/*
 * Returns the second angle of an Euler transformation in radians.
 */
Datum	spheretrans_theta(PG_FUNCTION_ARGS);

/*
 * Returns the third angle of an Euler transformation in radians.
 */
Datum	spheretrans_psi(PG_FUNCTION_ARGS);

/*
 * Returns the axis of an Euler transformation as three letter code.
 */
Datum	spheretrans_type(PG_FUNCTION_ARGS);

/*
 * Returns the Euler transformation (does nothing). This function is needed
 * for +strans operator.
 */
Datum	spheretrans(PG_FUNCTION_ARGS);

/*
 * Returns the inverse Euler transformation.
 */
Datum	spheretrans_invert(PG_FUNCTION_ARGS);

/*
 * Convert an Euler transformation to a ZXZ-axis transformation.
 */
Datum	spheretrans_zxz(PG_FUNCTION_ARGS);

/*
 * This function creates an Euler transformation from 3 angle values in
 * radians and three letter code used for axes. A letter can be X, Y or Z
 * (case-insensitive).
 */
Datum	spheretrans_from_float8_and_type(PG_FUNCTION_ARGS);

/*
 * Checks equality of two Euler transformations.
 */
Datum	spheretrans_equal(PG_FUNCTION_ARGS);

/*
 * Checks inequality of two Euler transformations.
 */
Datum	spheretrans_not_equal(PG_FUNCTION_ARGS);

/*
 * Transforms an Euler transformation.
 */
Datum	spheretrans_trans(PG_FUNCTION_ARGS);

/*
 * Transforms inverse an Euler transformations.
 */
Datum	spheretrans_trans_inv(PG_FUNCTION_ARGS);

/*
 * Transforms a spherical point.
 */
Datum	spheretrans_point(PG_FUNCTION_ARGS);

/*
 * Perform inverse transformation of a spherical point.
 */
Datum	spheretrans_point_inverse(PG_FUNCTION_ARGS);

#endif
