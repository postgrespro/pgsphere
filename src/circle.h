#ifndef __PGS_CIRCLE_H__
#define __PGS_CIRCLE_H__

#include "euler.h"

/* Spherical circle declarations */

/*
 * Spherical circle data structure: center and radius.
 */
typedef struct
{
	SPoint	center;		/* the center of circle */
	float8	radius;		/* the circle radius in radians */
} SCIRCLE;

/*
 * Checks whether two circles are equal.
 */
bool	scircle_eq(const SCIRCLE *c1, const SCIRCLE *c2);

/*
 * Checks whether a circle contains a point.
 */
bool	spoint_in_circle(const SPoint *p, const SCIRCLE *c);

/*
 * Transforms a circle using an Euler transformation.
 */
void	euler_scircle_trans(SCIRCLE *out, const SCIRCLE *in, const SEuler *se);

/*
 * Takes the input and stores it as a spherical circle.
 */
Datum	spherecircle_in(PG_FUNCTION_ARGS);

/*
 * Checks whether two circles are equal.
 */
Datum	spherecircle_equal(PG_FUNCTION_ARGS);

/*
 * Checks whether two circles are not equal.
 */
Datum	spherecircle_equal_neg(PG_FUNCTION_ARGS);

/*
 * Calculate the distance of two circles. If they overlap, this function
 * returns 0.0.
 */
Datum	spherecircle_distance(PG_FUNCTION_ARGS);

/*
 * Calculate the distance of a circle and a point. If a circle contains a point,
 * this function returns 0.0.
 */
Datum	spherecircle_point_distance(PG_FUNCTION_ARGS);

/*
 * Calculate the distance of a point and a circle. If a circle contains a point,
 * this function returns 0.0.
 */
Datum	spherecircle_point_distance_com(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle contains a point.
 */
Datum	spherepoint_in_circle(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle doesn't contain a point.
 */
Datum	spherepoint_in_circle_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle contains a point.
 */
Datum	spherepoint_in_circle_com(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle doesn't contain a point.
 */
Datum	spherepoint_in_circle_com_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle is contained by other circle.
 */
Datum	spherecircle_in_circle(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle is not contained by other circle.
 */
Datum	spherecircle_in_circle_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether a circle contains other circle.
 */
Datum	spherecircle_in_circle_com(PG_FUNCTION_ARGS);

/*
 * Checks whether circle does not contain other circle.
 */
Datum	spherecircle_in_circle_com_neg(PG_FUNCTION_ARGS);

/*
 * Checks whether two circles overlap.
 */
Datum	spherecircle_overlap(PG_FUNCTION_ARGS);

/*
 * Checks whether two circles overlap.
 */
Datum	spherecircle_overlap_neg(PG_FUNCTION_ARGS);

/*
 * Returns the center of a circle.
 */
Datum	spherecircle_center(PG_FUNCTION_ARGS);

/*
 * Returns the radius of a circle.
 */
Datum	spherecircle_radius(PG_FUNCTION_ARGS);

/*
 * Converts a point to a circle.
 */
Datum	spherepoint_to_circle(PG_FUNCTION_ARGS);

/*
 * Creates a circle from center and radius.
 */
Datum	spherecircle_by_center(PG_FUNCTION_ARGS);

/*
 * Calculates the area of a circle in square radians.
 */
Datum	spherecircle_area(PG_FUNCTION_ARGS);

/*
 * Calculates the circumference of a circle in radians.
 */
Datum	spherecircle_circ(PG_FUNCTION_ARGS);

/*
 * Transforms a circle using an Euler transformation.
 */
Datum	spheretrans_circle(PG_FUNCTION_ARGS);

/*
 * Inverse transformation of a circle using an Euler transformation.
 */
Datum	spheretrans_circle_inverse(PG_FUNCTION_ARGS);

#endif
