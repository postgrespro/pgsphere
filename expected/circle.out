\set ECHO none
 set_sphere_output_precision 
-----------------------------
 SET 8
(1 row)

-- Input/Output ---
SELECT set_sphere_output( 'RAD' );
 set_sphere_output 
-------------------
 SET RAD
(1 row)

SELECT '< (1h 2m 30s , +1d 2m 30s), 1.0d >'::scircle;
                  scircle                  
-------------------------------------------
 <(0.2727077 , 0.018180513) , 0.017453293>
(1 row)

SELECT '< (1h 2m 30s , +1d 2m 30s), 1d 30m >'::scircle;
                  scircle                  
-------------------------------------------
 <(0.2727077 , 0.018180513) , 0.026179939>
(1 row)

SELECT '< (1h 2m 30s , +1d 2m 30s), 0.1 >'::scircle;
              scircle              
-----------------------------------
 <(0.2727077 , 0.018180513) , 0.1>
(1 row)

SELECT '< (1h 2m 30s , +1d 2m 30s), 90d >'::scircle;
                 scircle                 
-----------------------------------------
 <(0.2727077 , 0.018180513) , 1.5707963>
(1 row)

SELECT set_sphere_output( 'DEG' );
 set_sphere_output 
-------------------
 SET DEG
(1 row)

SELECT '< (1h 2m 30s , +1d 2m 30s), 1.0d >'::scircle;
            scircle            
-------------------------------
 <(15.625d , 1.0416667d) , 1d>
(1 row)

SELECT '< (1h 2m 30s , +1d 2m 30s), 1d 30m >'::scircle;
             scircle             
---------------------------------
 <(15.625d , 1.0416667d) , 1.5d>
(1 row)

SELECT '< (1h 2m 30s , +1d 2m 30s), 0.1 >'::scircle;
               scircle                
--------------------------------------
 <(15.625d , 1.0416667d) , 5.729578d>
(1 row)

SELECT '< (1h 2m 30s , +1d 2m 30s), 90d >'::scircle;
            scircle             
--------------------------------
 <(15.625d , 1.0416667d) , 90d>
(1 row)

SELECT set_sphere_output( 'DMS' );
 set_sphere_output 
-------------------
 SET DMS
(1 row)

SELECT '< (1h 2m 30s , +1d 2m 30s), 1.0d >'::scircle;
                   scircle                    
----------------------------------------------
 <( 15d 37m 30s , + 1d  2m 30s) ,  1d  0m 0s>
(1 row)

SELECT '< (1h 2m 30s , +1d 2m 30s), 1d 30m >'::scircle;
                   scircle                    
----------------------------------------------
 <( 15d 37m 30s , + 1d  2m 30s) ,  1d 30m 0s>
(1 row)

SELECT '< (1h 2m 30s , +1d 2m 30s), 0.1 >'::scircle;
                       scircle                        
------------------------------------------------------
 <( 15d 37m 30s , + 1d  2m 30s) ,  5d 43m 46.480625s>
(1 row)

SELECT '< (1h 2m 30s , +1d 2m 30s), 90d >'::scircle;
                   scircle                    
----------------------------------------------
 <( 15d 37m 30s , + 1d  2m 30s) , 90d  0m 0s>
(1 row)

SELECT set_sphere_output( 'HMS' );
 set_sphere_output 
-------------------
 SET HMS
(1 row)

SELECT '< (1h 2m 30s , +1d 2m 30s), 1.0d >'::scircle;
                   scircle                    
----------------------------------------------
 <(  1h  2m 30s , + 1d  2m 30s) ,  1d  0m 0s>
(1 row)

SELECT '< (1h 2m 30s , +1d 2m 30s), 1d 30m >'::scircle;
                   scircle                    
----------------------------------------------
 <(  1h  2m 30s , + 1d  2m 30s) ,  1d 30m 0s>
(1 row)

SELECT '< (1h 2m 30s , +1d 2m 30s), 0.1 >'::scircle;
                       scircle                        
------------------------------------------------------
 <(  1h  2m 30s , + 1d  2m 30s) ,  5d 43m 46.480625s>
(1 row)

SELECT '< (1h 2m 30s , +1d 2m 30s), 90d >'::scircle;
                   scircle                    
----------------------------------------------
 <(  1h  2m 30s , + 1d  2m 30s) , 90d  0m 0s>
(1 row)

SELECT set_sphere_output( 'DMS' );
 set_sphere_output 
-------------------
 SET DMS
(1 row)

-- Functions
-- checking spherical circle operators
SELECT scircle '<(0, 90d),1>'   = '<(0, 90d),1>' ;
 ?column? 
----------
 t
(1 row)

SELECT scircle '<(0,-90d),1>'  <> '<(0, 90d),1>' ;
 ?column? 
----------
 t
(1 row)

-- float8 dist(scircle,scircle)
SELECT 180.0*dist('<( 0h 2m 30s , 10d 0m 0s), 0.1d>'::scircle,'<( 0h 2m 30s , -10d 0m 0s),0.1d>'::scircle)/pi();
 ?column? 
----------
     19.8
(1 row)

SELECT 180.0*dist('<( 0h 0m 00s , 0d 0m 0s), 1.0d>'::scircle,'<( 1h 0m 00s , 0d 0m 0s),1.0d>'::scircle)/pi();
 ?column? 
----------
       13
(1 row)

SELECT 180.0*dist('<( 23h 30m 00s , 0d 0m 0s), 1.0d>'::scircle,'<( 1h 0m 00s , 0d 0m 0s),1.5d>'::scircle)/pi();
 ?column? 
----------
       20
(1 row)

SELECT 180.0*dist('<( 0h 40m 00s , 0d 0m 0s), 1.0d>'::scircle,'<( 0h 50m 00s , 0d 0m 0s),1.0d>'::scircle)/pi();
 ?column? 
----------
      0.5
(1 row)

SELECT 180.0*dist('<( 0h 40m 00s , 0d 0m 0s), 1.5d>'::scircle,'<( 0h 50m 00s , 0d 0m 0s),1.5d>'::scircle)/pi();
 ?column? 
----------
        0
(1 row)

SELECT 180.0*dist('<( 0h 40m 00s , 90d 0m 0s), 1.5d>'::scircle,'<( 0h 50m 00s , 0d 0m 0s),1.5d>'::scircle)/pi();
 ?column? 
----------
       87
(1 row)

SELECT 180.0*dist('<( 0h 40m 00s , 90d 0m 0s), 1.0d>'::scircle,'<( 0h 50m 00s , -90d 0m 0s),1.0d>'::scircle)/pi();
 ?column? 
----------
      178
(1 row)

-- spoint center(scircle)
SELECT center('< (1h 2m 30s , +1d 2m 30s), 1.0d >'::scircle);
            center             
-------------------------------
 ( 15d 37m 30s , + 1d  2m 30s)
(1 row)

SELECT center('< (1h 2m 30s , +90d 0m 0s), 1.0d >'::scircle);
            center            
------------------------------
 ( 15d 37m 30s , +90d  0m 0s)
(1 row)

-- scircle scircle(spoint)
SELECT scircle('(0d,0d)'::spoint);
                  scircle                   
--------------------------------------------
 <(  0d  0m 0s , + 0d  0m 0s) ,  0d  0m 0s>
(1 row)

SELECT scircle('(0d,90d)'::spoint);
                  scircle                   
--------------------------------------------
 <(  0d  0m 0s , +90d  0m 0s) ,  0d  0m 0s>
(1 row)

SELECT scircle('(0d,-90d)'::spoint);
                  scircle                   
--------------------------------------------
 <(  0d  0m 0s , -90d  0m 0s) ,  0d  0m 0s>
(1 row)

-- Operators
-- = operator
-- should be "true"
SELECT '< (1h 0m 0s,+1d), 1.5d >'::scircle='< (1h 0m 0s,+1d), 1.5d >'::scircle ;
 ?column? 
----------
 t
(1 row)

SELECT '< (1h 0m 0s,+1d), 1.5d >'::scircle='< (25h 0m 0s,+1d), 1d 30m >'::scircle ;
 ?column? 
----------
 t
(1 row)

SELECT '< (1h 0m 0s,+95d), 1.5d >'::scircle='< (13h 0m 0s,+85d), 1.5d >'::scircle ;
 ?column? 
----------
 t
(1 row)

SELECT '< (1h 0m 0s,-95d), 1.5d >'::scircle='< (13h 0m 0s,-85d), 1.5d >'::scircle ;
 ?column? 
----------
 t
(1 row)

SELECT '< (1h 0m 0s,+90d), 1.5d >'::scircle='< (2h 0m 0s,+90d), 1.5d >'::scircle ;
 ?column? 
----------
 t
(1 row)

SELECT '< (1h 0m 0s,-90d), 1.5d >'::scircle='< (2h 0m 0s,-90d), 1.5d >'::scircle ;
 ?column? 
----------
 t
(1 row)

-- should be "false"
SELECT '< (1h 0m 0s,+1d), 1.5d >'::scircle='< (1h 0m 0s,-1d), 1.5d >'::scircle ;
 ?column? 
----------
 f
(1 row)

SELECT '< (1h 0m 0s,+1d), 1.5d >'::scircle='< (1h 0m 0s,+1d), 2.5d >'::scircle ;
 ?column? 
----------
 f
(1 row)

-- <> operator
-- should be "false"
SELECT '< (1h 0m 0s,+1d), 1.5d >'::scircle <> '< (1h 0m 0s,+1d), 1.5d >'::scircle ;
 ?column? 
----------
 f
(1 row)

SELECT '< (1h 0m 0s,+1d), 1.5d >'::scircle <> '< (25h 0m 0s,+1d), 1d 30m >'::scircle ;
 ?column? 
----------
 f
(1 row)

SELECT '< (1h 0m 0s,+95d), 1.5d >'::scircle <> '< (13h 0m 0s,+85d), 1.5d >'::scircle ;
 ?column? 
----------
 f
(1 row)

SELECT '< (1h 0m 0s,-95d), 1.5d >'::scircle <> '< (13h 0m 0s,-85d), 1.5d >'::scircle ;
 ?column? 
----------
 f
(1 row)

SELECT '< (1h 0m 0s,+90d), 1.5d >'::scircle <> '< (2h 0m 0s,+90d), 1.5d >'::scircle ;
 ?column? 
----------
 f
(1 row)

SELECT '< (1h 0m 0s,-90d), 1.5d >'::scircle <> '< (2h 0m 0s,-90d), 1.5d >'::scircle ;
 ?column? 
----------
 f
(1 row)

-- should be "true"
SELECT '< (1h 0m 0s,+1d), 1.5d >'::scircle <> '< (1h 0m 0s,-1d), 1.5d >'::scircle ;
 ?column? 
----------
 t
(1 row)

SELECT '< (1h 0m 0s,+1d), 1.5d >'::scircle <> '< (1h 0m 0s,+1d), 2.5d >'::scircle ;
 ?column? 
----------
 t
(1 row)

-- && operator
-- should be "true"
SELECT  '< (1h 0m 0s,+1d), 1.5d >'::scircle && '< (1h 0m 0s,+1d), 1.5d >'::scircle;
 ?column? 
----------
 t
(1 row)

SELECT  '< (1h 0m 0s,+1d), 1.5d >'::scircle && '< (1h 0m 0s,+1d), 0.5d >'::scircle;
 ?column? 
----------
 t
(1 row)

SELECT  '< (1h 0m 0s,+1d), 1.5d >'::scircle && '< (1h 0m 0s,-1d), 1.5d >'::scircle;
 ?column? 
----------
 t
(1 row)

SELECT  '< (1h 0m 0s,0d), 15d >'::scircle && '< (0h 0m 0s,0d), 15d >'::scircle;
 ?column? 
----------
 t
(1 row)

SELECT  '< (2h 0m 0s,0d), 15d >'::scircle && '< (0h 0m 0.1s,0d), 15d >'::scircle;
 ?column? 
----------
 t
(1 row)

SELECT  '< (1h 0m 0s,0d), 15d >'::scircle && '< (23h 0m 0.1s,0d), 15d >'::scircle;
 ?column? 
----------
 t
(1 row)

-- should be "false"
SELECT  '< (1h 0m 0s,+1d), 0.5d >'::scircle && '< (1h 0m 0s,-1d), 0.5d >'::scircle;
 ?column? 
----------
 f
(1 row)

SELECT  '< (1d 0m 0s,+1d), 1.0d >'::scircle && '< (0d 0m 0s,0d), 0.1d >'::scircle;
 ?column? 
----------
 f
(1 row)

-- @@ operator
SELECT @@ '< (1h 2m 3s , +1d 2m 3s), 1.0d >'::scircle;
           ?column?           
------------------------------
 ( 15d 30m 45s , + 1d  2m 3s)
(1 row)

SELECT @@ '< (1h 2m 3s , +90d 0m 0s), 1.0d >'::scircle;
           ?column?           
------------------------------
 ( 15d 30m 45s , +90d  0m 0s)
(1 row)

-- <-> operator
SELECT 180.0*('<( 0h 2m 30s , 10d 0m 0s), 0.1d>'::scircle<->'<( 0h 2m 30s , -10d 0m 0s),0.1d>'::scircle)/pi();
 ?column? 
----------
     19.8
(1 row)

SELECT 180.0*('<( 0h 0m 00s , 0d 0m 0s), 1.0d>'::scircle<->'<( 1h 0m 00s , 0d 0m 0s),1.0d>'::scircle)/pi();
 ?column? 
----------
       13
(1 row)

SELECT 180.0*('<( 23h 30m 00s , 0d 0m 0s), 1.0d>'::scircle<->'<( 1h 0m 00s , 0d 0m 0s),1.5d>'::scircle)/pi();
 ?column? 
----------
       20
(1 row)

SELECT 180.0*('<( 0h 40m 00s , 0d 0m 0s), 1.0d>'::scircle<->'<( 0h 50m 00s , 0d 0m 0s),1.0d>'::scircle)/pi();
 ?column? 
----------
      0.5
(1 row)

SELECT 180.0*('<( 0h 40m 00s , 0d 0m 0s), 1.5d>'::scircle<->'<( 0h 50m 00s , 0d 0m 0s),1.5d>'::scircle)/pi();
 ?column? 
----------
        0
(1 row)

SELECT 180.0*('<( 0h 40m 00s , 90d 0m 0s), 1.5d>'::scircle<->'<( 0h 50m 00s , 0d 0m 0s),1.5d>'::scircle)/pi();
 ?column? 
----------
       87
(1 row)

SELECT 180.0*('<( 0h 40m 00s , 90d 0m 0s), 1.0d>'::scircle<->'<( 0h 50m 00s , -90d 0m 0s),1.0d>'::scircle)/pi();
 ?column? 
----------
      178
(1 row)

-- scircle @ scircle operator
-- should be "true"
SELECT  '< (1h 0m 0s,+1d),  0.5d >'::scircle @ '< (1h 0m 0s,+1d),  1.5d >'::scircle;
 ?column? 
----------
 t
(1 row)

SELECT  '< (2d 0m 0s,+1d),  0.5d >'::scircle @ '< (1d 0m 0s,0d),   3.5d >'::scircle;
 ?column? 
----------
 t
(1 row)

SELECT  '< (1h 0m 0s,+89d), 0.5d >'::scircle @ '< (1h 0m 0s,+90d), 1.5d >'::scircle;
 ?column? 
----------
 t
(1 row)

SELECT  '< (1h 0m 0s,-89d), 0.5d >'::scircle @ '< (1h 0m 0s,-90d), 1.5d >'::scircle;
 ?column? 
----------
 t
(1 row)

-- should be "false"
SELECT  '< (1h 0m 0s,+1d),  0.5d >'::scircle @ '< (2h 0m 0s,+1d),  1.5d >'::scircle;
 ?column? 
----------
 f
(1 row)

SELECT  '< (2d 0m 0s,+1d),  1.5d >'::scircle @ '< (0d 0m 0s, 0d),  3.5d >'::scircle;
 ?column? 
----------
 f
(1 row)

SELECT  '< (1h 0m 0s,+89d), 0.5d >'::scircle @ '< (1h 0m 0s,+90d), 1.0d >'::scircle;
 ?column? 
----------
 f
(1 row)

SELECT  '< (1h 0m 0s,-89d), 0.5d >'::scircle @ '< (1h 0m 0s,-90d), 1.0d >'::scircle;
 ?column? 
----------
 f
(1 row)

-- scircle ~ scircle operator
-- should be "true"
SELECT '< (1h 0m 0s,+1d),  1.5d >'::scircle  ~ '< (1h 0m 0s,+1d),  0.5d >'::scircle;
 ?column? 
----------
 t
(1 row)

SELECT '< (1d 0m 0s,0d),   3.5d >'::scircle  ~ '< (2d 0m 0s,+1d),  0.5d >'::scircle;
 ?column? 
----------
 t
(1 row)

SELECT '< (1h 0m 0s,+90d), 1.5d >'::scircle  ~ '< (1h 0m 0s,+89d), 0.5d >'::scircle;
 ?column? 
----------
 t
(1 row)

SELECT '< (1h 0m 0s,-90d), 1.5d >'::scircle  ~ '< (1h 0m 0s,-89d), 0.5d >'::scircle;
 ?column? 
----------
 t
(1 row)

-- should be "false"
SELECT '< (2h 0m 0s,+1d),  1.5d >'::scircle  ~ '< (1h 0m 0s,+1d),  0.5d >'::scircle;
 ?column? 
----------
 f
(1 row)

SELECT '< (0d 0m 0s, 0d),  3.5d >'::scircle  ~ '< (2d 0m 0s,+1d),  1.5d >'::scircle;
 ?column? 
----------
 f
(1 row)

SELECT '< (1h 0m 0s,+90d), 1.0d >'::scircle  ~ '< (1h 0m 0s,+89d), 0.5d >'::scircle;
 ?column? 
----------
 f
(1 row)

SELECT '< (1h 0m 0s,-90d), 1.0d >'::scircle  ~ '< (1h 0m 0s,-89d), 0.5d >'::scircle;
 ?column? 
----------
 f
(1 row)

-- spoint @ scircle operator
-- should be "true"
SELECT  '(1h 0m 0s,+1d) '::spoint @ '< (1h 0m 0s,+1d),  1.5d >'::scircle;
 ?column? 
----------
 t
(1 row)

SELECT  '(2d 0m 0s,+1d) '::spoint @ '< (1d 0m 0s, 0d),  3.5d >'::scircle;
 ?column? 
----------
 t
(1 row)

SELECT  '(1h 0m 0s,+89d)'::spoint @ '< (1h 0m 0s,+90d), 1.5d >'::scircle;
 ?column? 
----------
 t
(1 row)

SELECT  '(1h 0m 0s,-89d)'::spoint @ '< (1h 0m 0s,-90d), 1.5d >'::scircle;
 ?column? 
----------
 t
(1 row)

SELECT  '(1h 0m 0s,+89d)'::spoint @ '< (1h 0m 0s,+90d), 1.0d >'::scircle;
 ?column? 
----------
 t
(1 row)

SELECT  '(1h 0m 0s,-89d)'::spoint @ '< (1h 0m 0s,-90d), 1.0d >'::scircle;
 ?column? 
----------
 t
(1 row)

-- should be "false"
SELECT  '(1h 0m  0s,+1d    )'::spoint @ '< (2h 0m 0s,-1d),  1.5d >'::scircle;
 ?column? 
----------
 f
(1 row)

SELECT  '(3d 30m 0s,+1d    )'::spoint @ '< (0d 0m 0s, 0d),  3.5d >'::scircle;
 ?column? 
----------
 f
(1 row)

SELECT  '(1h 0m  0s,+88.99d)'::spoint @ '< (1h 0m 0s,+90d), 1.0d >'::scircle;
 ?column? 
----------
 f
(1 row)

SELECT  '(1h 0m  0s,-88.99d)'::spoint @ '< (1h 0m 0s,-90d), 1.0d >'::scircle;
 ?column? 
----------
 f
(1 row)

-- spoint ~ scircle operator
-- should be "true"
SELECT '< (1h 0m 0s,+1d),  1.5d >'::scircle ~ '(1h 0m 0s,+1d) '::spoint;
 ?column? 
----------
 t
(1 row)

SELECT '< (1d 0m 0s, 0d),  3.5d >'::scircle ~ '(2d 0m 0s,+1d) '::spoint;
 ?column? 
----------
 t
(1 row)

SELECT '< (1h 0m 0s,+90d), 1.5d >'::scircle ~ '(1h 0m 0s,+89d)'::spoint;
 ?column? 
----------
 t
(1 row)

SELECT '< (1h 0m 0s,-90d), 1.5d >'::scircle ~ '(1h 0m 0s,-89d)'::spoint;
 ?column? 
----------
 t
(1 row)

SELECT '< (1h 0m 0s,+90d), 1.0d >'::scircle ~ '(1h 0m 0s,+89d)'::spoint;
 ?column? 
----------
 t
(1 row)

SELECT '< (1h 0m 0s,-90d), 1.0d >'::scircle ~ '(1h 0m 0s,-89d)'::spoint;
 ?column? 
----------
 t
(1 row)

-- should be "false"
SELECT '< (2h 0m 0s,-1d),  1.5d >'::scircle ~ '(1h 0m  0s,+1d    )'::spoint;
 ?column? 
----------
 f
(1 row)

SELECT '< (0d 0m 0s, 0d),  3.5d >'::scircle ~ '(3d 30m 0s,+1d    )'::spoint;
 ?column? 
----------
 f
(1 row)

SELECT '< (1h 0m 0s,+90d), 1.0d >'::scircle ~ '(1h 0m  0s,+88.99d)'::spoint;
 ?column? 
----------
 f
(1 row)

SELECT '< (1h 0m 0s,-90d), 1.0d >'::scircle ~ '(1h 0m  0s,-88.99d)'::spoint;
 ?column? 
----------
 f
(1 row)

